google/cloud/bigtable/read_rows_query.py

```diff
@@ -197,18 +241,135 @@ def add_range(
+        # each split_point defines last key that belongs to a segment
+        # our goal is to break up the query into subqueries that each operate in a single segment
+        split_points = [sample[0] for sample in shard_keys if sample[0]]
+
+        # handle row_keys
+        # use binary search to find the segment that each key belongs to
+        for this_key in list(self.row_keys):
+            # bisect_left: in case of exact match, pick left side (keys are inclusive ends)
+            segment_index = bisect_left(split_points, this_key)
+            sharded_queries[segment_index].add_key(this_key)
+
+        # handle row_ranges
+        for this_range in self.row_ranges:
+            # defer to _shard_range helper
+            for segment_index, added_range in self._shard_range(
+                this_range, split_points
+            ):
+                sharded_queries[segment_index].add_range(added_range)
+        # return list of queries ordered by segment index
+        # pull populated segments out of sharded_queries dict
+        keys = sorted(list(sharded_queries.keys()))
+        # return list of queries
+        return [sharded_queries[k] for k in keys]
+
+    @staticmethod
+    def _shard_range(
+        orig_range: RowRange, split_points: list[bytes]
+    ) -> list[tuple[int, RowRange]]:
+        """
+        Helper function for sharding row_range into subranges that fit into
+        segments of the key-space, determined by split_points
+
+        Args:
+          - orig_range: a row range to split
+          - split_points: a list of row keys that define the boundaries of segments.
+                each point represents the inclusive end of a segment
+        Returns:
+          - a list of tuples, containing a segment index and a new sub-range.
         """
-        raise NotImplementedError
+        # 1. find the index of the segment the start key belongs to
+        if orig_range.start is None:
+            # if range is open on the left, include first segment
+            start_segment = 0
+        else:
+            # use binary search to find the segment the start key belongs to
+            # bisect method determines how we break ties when the start key matches a split point
+            # if inclusive, bisect_left to the left segment, otherwise bisect_right
+            bisect = bisect_left if orig_range.start.is_inclusive else bisect_right
+            start_segment = bisect(split_points, orig_range.start.key)
+
+        # 2. find the index of the segment the end key belongs to
+        if orig_range.end is None:
+            # if range is open on the right, include final segment
+            end_segment = len(split_points)
+        else:
+            # use binary search to find the segment the end key belongs to.
+            # optimization: remove keys up to start_segment from searched list,
+            # then add start_segment back to get the correct index
```

Comment:
why not just pass `lo=start_segment` to bisect_left?

Reply:
good catch